feat: armonik domain definitions aep (#47)
aneoconsulting/ArmoniK.Community: AEP/aep-0000y.md

This AEP proposes a comprehensive definition of ArmoniK's domain. A domain is an area of interest or control within our system. Our goal is to clearly define the core domain entities to facilitate further development.

# Motivation

The purpose of this document is to establish a clear and consistent definition of ArmoniK's domain, enabling other documents and standards to effectively reuse these concepts.

# Rationale

To describe more complex features of ArmoniK accurately, it is crucial to have a shared vocabulary and avoid redundancy. While some definitions already exist in the ArmoniK glossary, this AEP seeks to foster community discussion and approval to ensure a unified understanding and usage of these terms.

# Specification

## Worker

User-developed containerized software capable of performing one or several user-implemented tasks. A worker is built on top of a uniformized runtime environment. A worker process input data, outputing more data from its calculation, as while as it can produce new tasks that might be handled by itself or by different workers. A worker embeds an implementation of the communication protocols with the scheduling agent.

## Scheduling agent

> Containerized software cohabiting with a worker, running a specific algorithm to determine which tasks its associated workers should perform, scheduling tasks on the workers and monitoring their execution. It also manages all interactions between the worker and the databases (retrieving/saving data, creating new tasks, etc.), as well as managing workers errors and retrying/resubmitting failed tasks when necessary. A scheduling agent, like a worker, exists within a single partition.

Author: Maybe we shouldn't use "pod" once we aim to go bare metal soon and it is mostly associated with k3s

## Partition

> Logical segmentation of the cluster's pool of machines to distribute workloads according to usage. This feature is provided and handled by ArmoniK.

Reviewer: As it may change soon I think it's better not to give this precision.

## Task

> An atomic computation that takes one or several inputs and outputs one or several results. A task is launched by a client or by another task and processed by a worker. In ArmoniK, tasks cannot communicate with each other. They can, however, depend on each other via their input/output data, known as data dependency. 

Reviewer 2: ```suggestion
An atomic computation that takes one or several inputs and outputs one or several results. A task is launched by a client or by another task and processed by a worker. In ArmoniK, a task cannot communicate with another one directly. They can, however, depend on each other via their input/output data, known as data dependency. 
```

A task is defined by:

- **TaskOptions**: Set of parameters specifying the execution conditions, for instance:
    - *PartitionId*: The Id of the partition where the task must be executed
    - *Maximum Duration*: The maximum duration of a task
    - *Max Retries*: The maximum number of retries 
    - *Priority*: The priority level of the task execution (the algorithm performs its best effort on tasks' priorities; however, it is possible that the priority won't always be respected)

- **Data Dependencies**: Input data for a given task that depends on another unique task. Data dependencies formalize dependencies between tasks.

> - **Expected Outputs**: Data that must be generated as output from a task. If a task submits new tasks then it can transfer responsibility for generating all or part of its outputs to the tasks it submits.

Reviewer 2: Above you are speaking about results, why not use this term here?

## Blob 

Abstraction for ArmoniK's task-related data, including data dependencies and expected outputs. 

> - **Blob Metadata**: Abstraction for referring to a blob, whether it exists or not. Blob metadata might be used when referring to data that is expected to exist in its full form in the future, such as expected outputs or data dependencies before the blob content is created.

Reviewer 2: ```suggestion
- **Blob Metadata**: Abstraction which refer to a blob, whether it exists or not. Blob metadata might be used when referring to data that is expected to exist in its full form in the futur, such as expected outputs or data dependencies before the blob content is created.
```

## Session

> A session is a logical container for tasks and associated data (task status, blobs, errors, etc.). Every task is submitted within a session. An existing session can be resumed to retrieve data or submit new tasks. When a session is canceled, all associated executions still in progress are interrupted.

Reviewer 2: ```suggestion
A session is a logical container for tasks and associated data (task status, blobs, errors, etc.). Every task is submitted within a session. An existing session can be paused and resumed to retrieve data or submit new tasks. When a session is canceled, all associated executions still in progress are interrupted.
```

# AEP : ArmoniK's Domain Definition

|                   |ArmoniK Enhancement Proposal|
---:                |:---
**AEP**             | 
**Title**           | Architecture for ArmoniK Extesion development
**Author**          | Italo Aguiar <<iaguiar@aneo.fr>>, Quentin Delamea <<qdelamea@aneo.fr>>
**Status**          | Draft
**Type**            | Standard
**Creation Date**   | 2024-06-04

# Abstract

This AEP proposes a comprehensive definition of ArmoniK's domain. A domain is an area of interest or control within our system. Our goal is to clearly define the core domain entities to facilitate further development.

# Motivation

The purpose of this document is to establish a clear and consistent definition of ArmoniK's domain, enabling other documents and standards to effectively reuse these concepts.

# Rationale

To describe more complex features of ArmoniK accurately, it is crucial to have a shared vocabulary and avoid redundancy. While some definitions already exist in the ArmoniK glossary, this AEP seeks to foster community discussion and approval to ensure a unified understanding and usage of these terms.

# Specification

## Worker

> User-developed containerized software capable of performing one or several tasks depending on its implementation. A worker can simply take input data and perform calculations on it to return a result. A worker can also submit new tasks that will be self-performed or handled by different workers or other instances of itself.

Reviewer: I think we need to be more specific:
- a worker can do both: return results and submit additionnal tasks;
- a worker embeds an implementation of the communication protocols with the scheduling agent (in fact a worker is an implementation inside a kind of shell);

## Scheduling agent

> Containerized software cohabiting with a worker within a pod, running a specific algorithm to determine which tasks "its" worker (the one with which it shares the pod) should perform. It also manages all interactions between the worker and the databases (retrieving/saving data, creating new tasks, etc.), as well as managing worker errors and retrying/resubmitting failed tasks when necessary. A scheduling agent, like a worker, exists within a single partition.

Reviewer: This definition should be clearer and more agnostic:
- ArmoniK will eventually have no need for Kubernetes so we should avoid terms like "pods";
- scheduling agent will eventually supervise several workers so we should state it;
- I might use more precise terms about the role of the scheduling agent: it schedules tasks on the workers, then monitor the execution => it performs a part of the global scheduling algorithm of ArmoniK.

## Partition

> Logical segmentation of the Kubernetes cluster's pool of machines to distribute workloads according to usage. This feature is provided and handled by ArmoniK.

Reviewer: Here again you should remove the reference to Kubernetes.

## Task

> An atomic computation that takes one or several inputs and outputs one or several results. A task is launched by a client and processed by a worker. In ArmoniK, tasks cannot communicate with each other. They can, however, depend on each other via their input/output data, known as data dependency. 

Reviewer: Tasks ability to submit other tasks is missing. Perhaps we could find a less abstract way of defining a task.

A task is defined by:

> - **TaskOptions**: Class that describes the options related to task execution, for instance:

Reviewer: ```suggestion
- **TaskOptions**: Set of parameters specifying the execution conditions, for instance:
```

>     - *PartitionId*: The Id of the partition where the task must be executed

Reviewer: ```suggestion
    - *Partitions*: The list of the partitions where the task can be executed
```

Author: When sending a taskOptions we only inform a PartitionId. Partitions are defined mostly when sessions are being created, but a task has one partition only.

    - *Maximum Duration*: The maximum duration of a task
    - *Max Retries*: The maximum number of retries 

>     - *Priority*: The priority level of the task execution

Reviewer: It may be a good idea to specify that this property is not always respected. The algorithm performs a best effort on tasks' priorities.

> - **TaskStatus**: Status of task processing

Reviewer: It doesn't belong to a task definition. Remove it.

> - **Payload**: Input data for a task that does not depend on any other task. This is the actual binary input of a task.

Reviewer: The payload will likely be remove in future versiosn of ArmoniK. I don't know if we should keep it here. Likewise, it is currently just a special data dependency.

Reviewer 3: I agree. We will remove it in the future. An alternative could be to pass the ID of the "payload" from the data dependencies as a custom field in the Task Options. Then, users can do whatever they want.
SDKs developpers can expose a payload if they think it is suitable but the underlying system will be more generic.

> - **Data Dependencies**: Input data for a given task that depends on another unique task. Data dependencies formalize dependencies between tasks.

Reviewer: ```suggestion
- **Data Dependencies**: Input data that depends on another unique task or is provided by users. Data dependencies formalize dependencies between tasks.
```

> - **Expected Outputs**: Data that must be generated as output from a task.

Reviewer: ```suggestion
- **Expected Outputs**: Data that must be generated as output from a task. If a task submits new tasks then it can transfer responsibility for generating all or part of its outputs to the tasks it submits.
```

## Blob 

> Abstraction for ArmoniK's task-related data, including payloads, data dependencies, and expected outputs. 

Reviewer: Here again I think we should remove the payload.
